chain/chain/src/types.rs

```diff
@@ -127,6 +132,119 @@ impl BlockHeaderInfo {
     }
 }
 
+/// Block economics config taken from genesis config
+pub struct BlockEconomicsConfig {
+    gas_price_adjustment_rate: Rational,
+    min_gas_price: Balance,
+    max_gas_price: Balance,
+    genesis_protocol_version: ProtocolVersion,
+}
+
+impl BlockEconomicsConfig {
+    /// Compute min gas price according to protocol version and chain id. We only upgrade gas price
+    /// for betanet, testnet and mainnet.
+    pub fn min_gas_price(&self, protocol_version: ProtocolVersion) -> Balance {
+        if self.genesis_protocol_version < protocol_version
+            && protocol_version >= MIN_PROTOCOL_VERSION_NEP_92
+        {
+            MIN_GAS_PRICE_NEP_92
+        } else {
+            self.min_gas_price
+        }
+    }
+
+    pub fn max_gas_price(&self, _protocol_version: ProtocolVersion) -> Balance {
+        self.max_gas_price
+    }
+
+    pub fn gas_price_adjustment_rate(&self, _protocol_version: ProtocolVersion) -> Rational {
+        self.gas_price_adjustment_rate
+    }
+}
+
+impl From<&ChainGenesis> for BlockEconomicsConfig {
+    fn from(chain_genesis: &ChainGenesis) -> Self {
+        BlockEconomicsConfig {
+            gas_price_adjustment_rate: chain_genesis.gas_price_adjustment_rate,
+            min_gas_price: chain_genesis.min_gas_price,
+            max_gas_price: chain_genesis.max_gas_price,
+            genesis_protocol_version: chain_genesis.protocol_version,
+        }
+    }
+}
+
+/// Chain genesis configuration.
+#[derive(Clone)]
+pub struct ChainGenesis {
+    pub time: DateTime<Utc>,
+    pub height: BlockHeight,
+    pub gas_limit: Gas,
+    pub min_gas_price: Balance,
+    pub max_gas_price: Balance,
+    pub total_supply: Balance,
+    pub max_inflation_rate: Rational,
+    pub gas_price_adjustment_rate: Rational,
+    pub transaction_validity_period: NumBlocks,
+    pub epoch_length: BlockHeightDelta,
+    pub protocol_version: ProtocolVersion,
+    pub chain_id: String,
+}
+
+impl ChainGenesis {
+    pub fn new(
+        time: DateTime<Utc>,
+        height: BlockHeight,
+        gas_limit: Gas,
+        min_gas_price: Balance,
+        max_gas_price: Balance,
+        total_supply: Balance,
+        max_inflation_rate: Rational,
+        gas_price_adjustment_rate: Rational,
+        transaction_validity_period: NumBlocks,
+        epoch_length: BlockHeightDelta,
+        protocol_version: ProtocolVersion,
+        chain_id: String,
+    ) -> Self {
+        Self {
+            time,
+            height,
+            gas_limit,
+            min_gas_price,
+            max_gas_price,
+            total_supply,
+            max_inflation_rate,
+            gas_price_adjustment_rate,
+            transaction_validity_period,
+            epoch_length,
+            protocol_version,
+            chain_id,
+        }
+    }
```

Comment:
Do we really need this? I mean, it may easily introduce some lurking bugs due to typos or missed order of the values. If we are fine with exposing all the fields as `pub`, I don't see a need in this method (at least the way it is currently implemented).

Reply:
Good point